includes/utils.php

```diff
@@ -1031,6 +1043,176 @@ function process_media( $url, $post_id, $args = [] ) {
+			! in_array( 'wp-image-' . $media_item['id'], $classes, true )
+		) {
+			continue;
+		}
+
+		// Try to determine the image size from the size class WordPress adds.
+		$image_size   = 'full';
+		$classes      = explode( ' ', $processor->get_attribute( 'class' ) ?? [] );
+		$size_classes = array_filter(
+			$classes,
+			function ( $image_class ) {
+				return false !== strpos( $image_class, 'size-' );
+			}
+		);
+
+		if ( ! empty( $size_classes ) ) {
+			// If an image happens to have multiple size classes, just use the first.
+			$size_class = reset( $size_classes );
+			$image_size = str_replace( 'size-', '', $size_class );
+		}
+
+		$src = wp_get_attachment_image_url( $image_id, $image_size );
+
+		// If the image size can't be found, try to get the full size.
+		if ( ! $src ) {
+			$src = wp_get_attachment_image_url( $image_id, 'full' );
+
+			// If we still don't have an image, skip this block.
+			if ( ! $src ) {
+				continue;
+			}
+		}
+
+		$processor->set_attribute( 'src', $src );
+		$processor->add_class( 'wp-image-' . $image_id );
+		$processor->remove_class( 'wp-image-' . $media_item['id'] );
+		$processor->remove_attribute( 'srcset' );
+		$processor->remove_attribute( 'sizes' );
+	}
+
+	return $processor->get_updated_html();
+}
+
+/**
+ * Find and update an image block.
+ *
+ * @param array $blocks All blocks in a post.
+ * @param array $media_item The old media item details.
+ * @param int   $image_id The new image ID.
+ * @return array
+ */
+function update_image_block( array $blocks, array $media_item, int $image_id ) {
+	// Find and update all image blocks that match the old image ID.
+	foreach ( $blocks as $key => $block ) {
+		// Recurse into inner blocks.
+		if ( ! empty( $block['innerBlocks'] ) ) {
+			$blocks[ $key ]['innerBlocks'] = update_image_block( $block['innerBlocks'], $media_item, $image_id );
+		}
+
+		// If the block is an image block and the ID matches, update the ID and URL.
+		if ( 'core/image' === $block['blockName'] && $media_item['id'] === $block['attrs']['id'] ) {
+			$image_size = $block['attrs']['sizeSlug'] ?? 'full';
+
+			$blocks[ $key ]['attrs']['id'] = $image_id;
+
+			$processor = new \WP_HTML_Tag_Processor( $blocks[ $key ]['innerHTML'] );
+
+			// Use the HTML API to update the image src and class.
+			if ( $processor->next_tag( 'img' ) ) {
+				$src = wp_get_attachment_image_url( $image_id, $image_size );
+
+				// If the image size can't be found, try to get the full size.
+				if ( ! $src ) {
+					$src = wp_get_attachment_image_url( $image_id, 'full' );
+
+					// If we still don't have an image, skip this block.
+					if ( ! $src ) {
+						continue;
+					}
+				}
+
+				$processor->set_attribute( 'src', $src );
+				$processor->add_class( 'wp-image-' . $image_id );
+				$processor->remove_class( 'wp-image-' . $media_item['id'] );
+
+				$blocks[ $key ]['innerHTML']       = $processor->get_updated_html();
+				$blocks[ $key ]['innerContent'][0] = $processor->get_updated_html();
```

Comment:
Would love feedback on if there's a better approach here. I was hoping I could just update the block attributes and then re-render the block to have it rebuild the saved block markup but nothing I tried actually worked, so I ended up with this approach of basically doing a search/replace from the old URL to the new URL

Reply:
I tried using `render_block` and `serialize_block` to avoid directly assigning to `innerHTML`, but that broke the image block due to inconsistency between the content returned by `save()` and `edit()`.